src/RamIndexAnalysis.h

```diff
@@ -34,16 +34,160 @@
+        assert(bits.size() == other.bits.size());
+        std::transform(other.bits.begin(), other.bits.end(), bits.begin(), bits.begin(), std::bit_or<bool>());
+        return *this;
+    }
+    inline SearchSignature& operator&=(const SearchSignature& other) {
+        assert(bits.size() == other.bits.size());
+        std::transform(
+                other.bits.begin(), other.bits.end(), bits.begin(), bits.begin(), std::bit_and<bool>());
+        return *this;
+    }
+    inline SearchSignature& operator^=(const SearchSignature& other) {
+        assert(bits.size() == other.bits.size());
+        std::transform(
+                other.bits.begin(), other.bits.end(), bits.begin(), bits.begin(), std::bit_xor<bool>());
+        return *this;
+    }
+    inline SearchSignature& operator-=(const SearchSignature& other) {
+        assert(bits.size() == other.bits.size());
+        for (size_t i = 0; i < bits.size(); ++i) {
+            bits[i] = bits[i] && !other.bits[i];
+        }
+        return *this;
+    }
+    inline SearchSignature& operator<<=(const size_t n) {
+        size_t original_size = bits.size();
+        bits.insert(bits.begin(), n, false);
+        bits.erase(bits.end() - n, bits.end());
+        assert(bits.size() == original_size);
+        return *this;
+    }
+    inline SearchSignature& operator>>=(const size_t n) {
+        size_t original_size = bits.size();
+        bits.resize(bits.size() + n, false);
+        bits.erase(bits.begin(), bits.begin() + n);
+        assert(bits.size() == original_size);
+        return *this;
+    }
+    // set a bit
+    inline SearchSignature& set(size_t pos, bool val = true) {
+        assert(pos < bits.size());
+        bits[pos] = val;
+        return *this;
+    }
+
+    // bitwise operators
+    inline SearchSignature operator~() const {
+        SearchSignature s(*this);
+        s.bits.flip();
+        return s;
+    }
+    inline SearchSignature operator|(const SearchSignature& other) const {
+        SearchSignature s(*this);
+        return s |= other;
+    }
+    inline SearchSignature operator&(const SearchSignature& other) const {
+        SearchSignature s(*this);
+        return s &= other;
+    }
+    inline SearchSignature operator^(const SearchSignature& other) const {
+        SearchSignature s(*this);
+        return s ^= other;
+    }
+    inline SearchSignature operator-(const SearchSignature& other) const {
+        SearchSignature s(*this);
+        return s -= other;
+    }
+    inline SearchSignature operator<<(const size_t n) const {
+        SearchSignature s(*this);
+        return s <<= n;
+    }
+    inline SearchSignature operator>>(const size_t n) const {
+        SearchSignature s(*this);
+        return s >>= n;
+    }
+
+    inline size_t arity() const {
+        return bits.size();
+    }
+    friend std::ostream& operator<<(std::ostream& out, const SearchSignature& signature);
+
+private:
+    std::vector<bool> bits;
```

Comment:
`std::vector<bool>` is no longer required to be a template specialization in C++17, it is impl defined. If you want an actual bitset, you'll have to hand roll this yourself. Trivial enough, use a `uint32_t` as an element type and you'll likely get the small-vector-optimization to kick in (if impl in your STL).